sv.c

```diff
@@ -2098,9 +2098,10 @@ S_sv_2iuv_common(pTHX_ SV *const sv)
             SvIV_set(sv, I_V(SvNVX(sv)));
             if (SvNVX(sv) == (NV) SvIVX(sv)
 #ifndef NV_PRESERVES_UV
-                && SvIVX(sv) != IV_MIN /* avoid negating IV_MIN below */
-                && (((UV)1 << NV_PRESERVES_UV_BITS) >
-                    (UV)(SvIVX(sv) > 0 ? SvIVX(sv) : -SvIVX(sv)))
+                /* Optimizing compilers might merge two comparisons below
+                   into single comparison */
+                && -(IV)((UV)1 << NV_PRESERVES_UV_BITS) < SvIVX(sv)
```

Comment:
If `(IV)(1 << NV_PRESERVES_UV_BITS)` ends up being `IV_MIN`, this code will try to negate `IV_MIN` again. Actually, the real problem is one step before that: If `(UV)1 << NV_PRESERVES_UV_BITS` exceeds the range of `IV`, the result of the `IV` cast is implementation defined (or might raise a signal).
I believe this would happen if there is a platform where `NV_PRESERVES_UV_BITS` is equal to `IVSIZE * CHAR_BIT - 1`.

Reply:
> If `(UV)1 << NV_PRESERVES_UV_BITS` exceeds the range of `IV`,
[...]
> I believe this would happen if there is a platform where `NV_PRESERVES_UV_BITS` is equal to `IVSIZE * CHAR_BIT - 1`.

I think these would practically never happen (there seems to be no combination of such floating-point format and IV/UV format on platforms supported by perl), and even if that is the case, [static assertion in `sv.c`](https://github.com/Perl/perl5/blob/0e9cc8e5aaebeef20004deb2c6fdf911e60ba482/sv.c#L2292-L2293) will fail anyway.

Reply:
Maybe:
```
        && (IV)(((UV)1 << NV_PRESERVES_UV_BITS)-1) <= SvIVX(sv)
        && SvIVX(sv) <= (IV)-(((UV)1 << NV_PRESERVES_UV_BITS)-1)
```
instead, which avoids negating `IV_MIN`.

Defining symbols for the limits makes this a lot more readable:
```
#define MAX_NV_PRESERVED (IV)(((UV)1 << NV_PRESERVES_UV_BITS)-1)
#define MIN_NV_PRESERVED (IV)-(((UV)1 << NV_PRESERVES_UV_BITS)-1)

...
        && MIN_NV_PRESERVED <= SvIVX(sv)
        && SvIVX(sv) <= MAX_NV_PRESERVED
```
See https://godbolt.org/z/ThGbvYav4 for some generated code comparisons.

Reply:
That makes sense.  I will modify this PR.